manager/recurringjob.go

```diff
@@ -114,9 +114,64 @@ func (m *VolumeManager) UpdateRecurringJob(spec longhorn.RecurringJobSpec) (*lon
 }
 
 func (m *VolumeManager) DeleteRecurringJob(name string) error {
+	if err := m.CleanVolumeRecurringJob(name); err != nil {
+		return err
+	}
+
 	if err := m.ds.DeleteRecurringJob(name); err != nil {
 		return err
 	}
 	logrus.Infof("Deleted recurring job %v", name)
 	return nil
 }
+
+func (m *VolumeManager) CleanVolumeRecurringJob(name string) error {
+	recurringJob, err := m.ds.GetRecurringJob(name)
+	if err != nil {
+		return err
+	}
+
+	// check if other recurring jobs contains same groups
+	recurringJobs, err := m.ds.ListRecurringJobs()
+	if err != nil {
+		return err
+	}
+	rmGroups := []string{}
+	for _, group := range recurringJob.Spec.Groups {
+		inUse := 0
+		for _, obj := range recurringJobs {
+			if util.Contains(obj.Spec.Groups, group) {
+				inUse++
+			}
+			if inUse > 1 {
+				continue
+			}
+		}
+		if inUse == 1 {
+			rmGroups = append(rmGroups, group)
+		}
+	}
+
+	// delete volume labels
+	volumes, err := m.ds.ListVolumes()
+	if err != nil {
+		return err
+	}
+	for volName, vol := range volumes {
+		list, err := m.ListVolumeRecurringJob(volName)
+		if err != nil {
+			return err
+		}
+		if vol.Spec.LastAttachedBy == name && vol.Status.State != longhorn.VolumeStateDetached {
+			m.Detach(volName, "")
+		}
```

Comment:
Why we need this? I thought recurring job management should not impact volume management.

Reply:
Actually, recurring job also manages auto-attachment [here](https://github.com/longhorn/longhorn-manager/blob/master/app/recurring_job.go#L330). This was added in the case when the pod gets removed before it detaches the volume.

But since the controller now does the clean-up, this is not needed. Removed.